File: lib/stringlib/stringlib.go
package stringlib

import (
	"strings"

	"github.com/arnodel/golua/lib/packagelib"
	"github.com/arnodel/golua/luastrings"
	rt "github.com/arnodel/golua/runtime"
)

// LibLoader specifies how to load the string lib
var LibLoader = packagelib.Loader{
	Load: load,
	Name: "string",
}

func load(r *rt.Runtime) (rt.Value, func()) {
	pkg := rt.NewTable()
	pkgVal := rt.TableValue(pkg)

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(pkg, "byte", bytef, 3, false),
		r.SetEnvGoFunc(pkg, "char", char, 0, true),
		r.SetEnvGoFunc(pkg, "dump", dump, 2, false),
		r.SetEnvGoFunc(pkg, "find", find, 4, false),
		r.SetEnvGoFunc(pkg, "gmatch", gmatch, 3, false),
		r.SetEnvGoFunc(pkg, "gsub", gsub, 4, false),
		r.SetEnvGoFunc(pkg, "len", lenf, 1, false),
		r.SetEnvGoFunc(pkg, "lower", lower, 1, false),
		r.SetEnvGoFunc(pkg, "match", match, 3, false),
		r.SetEnvGoFunc(pkg, "upper", upper, 1, false),
		r.SetEnvGoFunc(pkg, "rep", rep, 3, false),
		r.SetEnvGoFunc(pkg, "reverse", reverse, 1, false),
		r.SetEnvGoFunc(pkg, "sub", sub, 3, false),
		r.SetEnvGoFunc(pkg, "format", format, 1, true),
		r.SetEnvGoFunc(pkg, "pack", pack, 1, true),
		r.SetEnvGoFunc(pkg, "packsize", packsize, 1, false),
		r.SetEnvGoFunc(pkg, "unpack", unpack, 3, false),
	)

	stringMeta := rt.NewTable()
	r.SetEnv(stringMeta, "__index", pkgVal)

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(stringMeta, "__add", string__add, 2, false),
		r.SetEnvGoFunc(stringMeta, "__sub", string__sub, 2, false),
		r.SetEnvGoFunc(stringMeta, "__mul", string__mul, 2, false),
		r.SetEnvGoFunc(stringMeta, "__div", string__div, 2, false),
		r.SetEnvGoFunc(stringMeta, "__idiv", string__idiv, 2, false),
		r.SetEnvGoFunc(stringMeta, "__mod", string__mod, 2, false),
		r.SetEnvGoFunc(stringMeta, "__pow", string__pow, 2, false),
		r.SetEnvGoFunc(stringMeta, "__unm", string__unm, 1, false),
	)
	r.SetStringMeta(stringMeta)

	return pkgVal, nil
}

func maxpos(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func minpos(i, j int) int {
	if i < j {
		return i
	}
	return j
}

func bytef(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	i, j := 1, 1
	if c.NArgs() >= 2 {
		ii, err := c.IntArg(1)
		if err != nil {
			return nil, err
		}
		i = luastrings.StringNormPos(s, int(ii))
		j = i
	}
	if c.NArgs() >= 3 {
		jj, err := c.IntArg(2)
		if err != nil {
			return nil, err
		}
		j = luastrings.StringNormPos(s, int(jj))
	}
	next := c.Next()
	i = maxpos(1, i)
	j = minpos(len(s), j)
	for i <= j {
		t.Push1(next, rt.IntValue(int64(s[i-1])))
		i++
	}
	return next, nil
}

func char(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	vals := c.Etc()
	buf := make([]byte, len(vals))
	for i, v := range vals {
		x, ok := rt.ToInt(v)
		if !ok {
			return nil, rt.NewErrorS("arguments must be integers")
		}
		if x < 0 || x > 255 {
			return nil, rt.NewErrorF("#%d out of range", i+1)
		}
		buf[i] = byte(x)
	}
	t.RequireBytes(len(buf))
	return c.PushingNext1(t.Runtime, rt.StringValue(string(buf))), nil
}

func lenf(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	return c.PushingNext1(t.Runtime, rt.IntValue(int64(len(s)))), nil
}

func lower(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	t.RequireBytes(len(s))
	s = strings.ToLower(string(s))
	return c.PushingNext1(t.Runtime, rt.StringValue(s)), nil
}

func upper(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	t.RequireBytes(len(s))
	s = strings.ToUpper(string(s))
	return c.PushingNext1(t.Runtime, rt.StringValue(s)), nil
}

func rep(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	ls, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	ln, err := c.IntArg(1)
	if err != nil {
		return nil, err
	}
	n := int(ln)
	if n < 0 {
		return nil, rt.NewErrorS("#2 out of range")
	}
	var sep []byte
	if c.NArgs() >= 3 {
		lsep, err := c.StringArg(2)
		if err != nil {
			return nil, err
		}
		sep = []byte(lsep)
	}
	if n == 0 {
		return c.PushingNext1(t.Runtime, rt.StringValue("")), nil
	}
	if n == 1 {
		return c.PushingNext1(t.Runtime, rt.StringValue(ls)), nil
	}
	if sep == nil {
		if len(ls)*n/n != len(ls) {
			// Overflow
			return nil, rt.NewErrorS("rep causes overflow")
		}
		t.RequireBytes(n * len(ls))
		return c.PushingNext1(t.Runtime, rt.StringValue(strings.Repeat(string(ls), n))), nil
	}
	s := []byte(ls)
	builder := strings.Builder{}
	sz1 := n * len(s)
	sz2 := (n - 1) * len(sep)
	sz := sz1 + sz2
	if sz1/n != len(s) || sz2/(n-1) != len(sep) || sz < 0 {
		return nil, rt.NewErrorS("rep causes overflow")
	}
	t.RequireBytes(n*len(s) + (n-1)*len(sep))
	builder.Grow(sz)
	builder.Write(s)
	for {
		n--
		if n == 0 {
			break
		}
		builder.Write(sep)
		builder.Write(s)
	}
	return c.PushingNext1(t.Runtime, rt.StringValue(builder.String())), nil
}

func reverse(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	t.RequireBytes(len(s))
	sb := []byte(s)
	l := len(s) - 1
	for i := 0; 2*i <= l; i++ {
		sb[i], sb[l-i] = sb[l-i], sb[i]
	}
	return c.PushingNext1(t.Runtime, rt.StringValue(string(sb))), nil
}

func sub(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	s, err := c.StringArg(0)
	if err != nil {
		return nil, err
	}
	ii, err := c.IntArg(1)
	if err != nil {
		return nil, err
	}
	i := luastrings.StringNormPos(s, int(ii))
	j := len(s)
	if c.NArgs() >= 3 {
		jj, err := c.IntArg(2)
		if err != nil {
			return nil, err
		}
		j = luastrings.StringNormPos(s, int(jj))
	}
	var slice string
	i = maxpos(1, i)
	j = minpos(len(s), j)
	if i <= len(s) && i <= j {
		t.RequireBytes(j - i + 1)
		slice = s[i-1 : j]
	}
	return c.PushingNext1(t.Runtime, rt.StringValue(slice)), nil
}

File: lib/stringlib/lua/stringlib.lua
local function errtest(f)
    return function(...)
        local ok, err = pcall(f, ...)
        if ok then
            print"OK"
        else
            print(err)
        end
    end
end

do
    local s = "hello"
    print(string.byte(s))
    --> =104

    print(string.byte(s, -1))
    --> =111

    print(string.byte(s, 2, 4))
    --> =101	108	108

    print(string.byte(s, -5, 2))
    --> =104	101
    
    print(string.byte(s, -2, -1))
    --> =108	111

    -- Byte can also be called as a method on strings
    print(s:byte(3))
    --> =108

    local errbyte = errtest(string.byte)

    errbyte()
    --> ~value needed

    errbyte({})
    --> ~must be a string

    errbyte("xxx", true)
    --> ~must be an integer

    errbyte("xxx", 1, nil)
    --> ~must be an integer
end

do
    print(string.char(65, 66, 67))
    --> =ABC

    local errchar = errtest(string.char)

    errchar(-1)
    --> ~out of range

    errchar(256)
    --> ~out of range

    errchar(1, 2, "x")
    --> ~must be integers
end

do
    print(string.len("abc"), string.len(""))
    --> =3	0

    errtest(string.len)()
    --> ~value needed

    errtest(string.len)(123)
    --> ~must be a string
end

do
    local s = "ABCdef123"
    print(s:lower())
    --> =abcdef123

    print(s:upper())
    --> =ABCDEF123

    errtest(string.lower)()
    --> ~value needed

    errtest(string.lower)(123)
    --> ~must be a string

    errtest(string.upper)()
    --> ~value needed

    errtest(string.upper)(123)
    --> ~must be a string
end

do
    local s = "xy"
    for i = 0, 3 do
        print(s:rep(i))
    end
    --> =
    --> =xy
    --> =xyxy
    --> =xyxyxy

    for i = 0, 3 do
        print(s:rep(i, "--"))
    end
    --> =
    --> =xy
    --> =xy--xy
    --> =xy--xy--xy

    local errrep = errtest(string.rep)

    errrep()
    --> ~2 arguments needed

    errrep(1, 2)
    --> ~must be a string

    errrep("xx", {})
    --> ~must be an integer

    errrep("xx", -1)
    --> ~out of range

    errrep("x", 2, 1)
    --> ~must be a string

    errrep("xxxxxxxxxx", 1 << 62)
    --> ~overflow

    errrep("xx", 1 << 62, ";")
    --> ~overflow
end

do
    local s = "EGASSEM TERCES"
    print(s:reverse())
    --> =SECRET MESSAGE

    print(string.reverse("12345"))
    --> =54321

    errtest(string.reverse)()
    --> ~value needed

    errtest(string.reverse)(123)
    --> ~must be a string
end

do
    local s = "abc"
    for i = -4, 4 do
        print(s:sub(i))
    end
    --> =abc
    --> =abc
    --> =bc
    --> =c
    --> =abc
    --> =abc
    --> =bc
    --> =c
    --> =

    print(s:sub(2, 3))
    --> =bc

    print(s:sub(3, 6))
    --> =c

    local subtest = errtest(string.sub)

    subtest()
    --> ~2 arguments needed

    subtest(1, 2)
    --> ~must be a string

    subtest("x", {})
    --> ~must be an integer

    subtest("xxx", 1, true)
    --> ~must be an integer
end

do
    local function pf(...)
        print(string.format(...))
    end
    
    local errf = errtest(string.format)

    pf("%s=%f", "pi", 3.14)
    --> =pi=3.140000

    pf("-%s-%s-%s", nil, true, false)
    --> =-nil-true-false

    pf("%% %q %%", [["hello"	123]])
    --> =% "\"hello\"\t123" %

    pf("%d//%5d//%-5d//%+d//%05d", 10.0, "10", 10, 10, 10)
    --> =10//   10//10   //+10//00010

    pf("%.2f~~%5.2f~~%-5.2f~~%+.2f~~%05.2f", 3.14, "3.14", 3.14, 3.14, 3.14)
    --> =3.14~~ 3.14~~3.14 ~~+3.14~~03.14

    -- To many values is OK
    pf("%s", 1, 2, 3)
    --> =1

    pf("%c", 65)
    --> =A

    errf("%c")
    --> ~not enough values

    errf("%c", {})
    --> ~invalid value

    errf("%d", "hello")
    --> ~invalid value

    pf("-%u-", 55)
    --> =-55-

    pf("%i", -12)
    --> =-12

    pf("%x", 255)
    --> =ff

    errf("%e")
    --> ~not enough values

    errf("%e", false)
    --> ~invalid value

    errf('"%s"')
    --> ~not enough values

    local t = {}
    setmetatable(t, {__tostring=function() error("bad", 0) end})
    errf("%s", t)
    --> =bad

    errf("%q")
    --> ~not enough values

    pf("[%q]", nil)
    --> =[nil]

    errf("%q", {})
    --> ~no literal

    pf("%q %q %q", false, 1, 1.5)
    --> =false 1 1.5

    errf("%t")
    --> ~not enough values

    pf("This is %t", true)
    --> =This is true

    errf("%t", 1)
    --> ~invalid value

    errf("%z")
    --> ~invalid format string

    -- Not enough values is not OK
    print(pcall(pf, "%s %d", 1))
    --> ~^false\t.*$

    errf()
    --> ~value needed

    errf(321)
    --> ~must be a string

end

do
    local function dl(...)
        return load(string.dump(...))
    end
    
    local function f(x)
        print(string.format("%s squared is %s", x, x*x))
    end

    print(type(string.dump(f)))
    --> =string

    dl(f)(2)
    --> =2 squared is 4

    dl(f, true)(10)
    --> =10 squared is 100

    local function g(x)
        return function(y)
            print("Working...")
            return x + y + 2
        end
    end

    print(dl(g)(3)(5))
    --> =Working...
    --> =10

    local errd = errtest(string.dump)

    errd()
    --> ~value needed

    errd(true)
    --> ~must be a lua function
end
